spi-proposal.md

```diff
@@ -0,0 +1,200 @@
+SPI
+===
+
+A proposal for a SPI protocol for Firmata.
+
+SPI is tricky to add to Firmata in a generic way since it is a fairly loose
+standard. There are variations in number of bits written and read, how the CS
+pin is used (if at all), use of other pins, etc. This proposal attempts to
+accommodate uses cases beyond the common:
+
+1. set cs pin LOW
+2. write (and optionally read) 1 or more bytes
+3. set cs Pin HIGH
+4. return data if read
+
+Each SPI device is initialized via a config message. This message sets the bit
+order data mode clock speed and optional CS pin for the device. Arrays of bytes
+can be transferred to and received from the slave device. If a CS pin is
+specified, it is pulled low at the beginning of a transfer and released after
+the last byte sent. However, the user can change the active CS pin state (HIGH
+to LOW rather than the default LOW to HIGH behavior) or even specify if the
+pin should be toggled between every byte vs the beginning and end of every
+sequence of bytes.
+
+A **sequence** byte is provide to allow breaking up a transfer into multiple
+packets or even enabling a continuous data stream (provided HW doesn't have
+tight timing requirements).
+
+There are 3 ways to send and receive data from the SPI slave device:
+
+1. Transfer - for each byte sent a byte is received
+2. Write - only write data (ignore any data retuned by the slave device)
+3. Read - only read data, writing 0 for each byte to be read
+
+
+## Protocol details
+
+### Begin
+
+Call once to initialize SPI hardware. Can optionally be set internally upon
+receiving first `SPI_CONFIG_DEVICE` message.
+
+```
+0:  START_SYSEX
+1:  SPI_DATA
+2:  SPI_BEGIN
+3:  channel (HW supports multiple SPI ports. range = 0-3, default = 0)
+4:  END_SYSEX
+```
+
+### Config
+
+Send once for each SPI device, specifying a new deviceId for each device
+(beginning from 0).
+
+Bit order: LSBFIRST = 0, MSBFIRST = 1
+
+Data modes:
+
+| mode    | clock polarity (CPOL) | clock phase (CPHA) |
+| --------|-----------------------|--------------------|
+| 0       | 0                     | 0                  |
+| 1       | 0                     | 1                  |
+| 2       | 1                     | 0                  |
+| 3       | 1                     | 1                  |
+
+
+```
+0:  START_SYSEX
+1:  SPI_DATA
+2:  SPI_CONFIG_DEVICE
+3:  deviceId | channel (bits 2-6: deviceId, bits 0-1: channel)
+4:  bitOrder (bit 0) | dataMode (bits 1-2, values 0-3) see mode column in table above
+5:  clockSpeed (bits 0 - 6)
+6:  clockSpeed (bits 7 - 14)
+7:  clockSpeed (bits 15 - 21)
+8:  clockSpeed (bits 22 - 28)
+9:  clockSpeed (bits 29 - 32)
+10: csPin (0-127) [optional] use csPin to identify individual devices
+11: csOptions (used only if csPin is set):
```

Comment:
These options (along with others potentially - depending on how many uses cases we want to try to support) may need to be specified in the REQUEST messages rather than in the CONFIG message. Need to investigate this further.